### PR TITLE
dma: Use correct len type

### PR DESCRIPTION
**Pull request by swapnili**
vfio_iommu_type1_dirty_bitmap_get.size is of type __u64
dma_controller_dirty_page_get() recieves it as int.

Fixes: #477

Signed-off-by: Swapnil Ingle <swapnil.ingle@nutanix.com>

### CODE REVIEW

File: include/libvfio-user.h [hunk 1]
```diff
@@ -66,7 +66,7 @@ typedef void *vfu_dma_addr_t;
 typedef struct {
     vfu_dma_addr_t dma_addr;
     int region;
-    int length;
+    uint64_t length;
```
Comment (jlevon) [inline include/libvfio-user.h:69]: but dma_init_sg(), and various friends, still take a 32-bit value, so I'm not sure how this works.

Also, we'd need to do something about vfu_dma_read / vfu_dma_write() - right now they are truncating a potentially 64-bit sg->length

**Review by tmakatos — CHANGES_REQUESTED** [review]
IMO there should be a unit test where the bug triggers without the fix and then a commit that fixes the bug. The SPDK log I've attached contains the values that in the dirty pages message, so it should be trivially to repro.

File: include/libvfio-user.h (hunk 1 above)
Comment (swapnili) [reply include/libvfio-user.h:69]: Shouldn't it be safe to put 32bit(unsigned) value into 64bit(unsigned)?
Or should we also change dma_init_sg() and friends to receive length in uint64_t

For vfu_dma_read / vfu_dma_write() right now the protocol expects 'count' to be uint32_t. How about having value check for <UINT32_MAX?

**Comment by swapnili** [comment]
> IMO there should be a unit test where the bug triggers without the fix and then a commit that fixes the bug. The SPDK log I've attached contains the values that in the dirty pages message, so it should be trivially to repro.

Sure I will try to repro it with Unit test

File: include/libvfio-user.h (hunk 1 above)
Comment (jlevon) [reply include/libvfio-user.h:69]: Was the original bug from using int rather uint32_t ?
You made a change such that dma_controller_dirty_page_get() takes a uint64_t, but without changing dma_init_sg() etc. you're dropping the top 32 bits on the floor

**Comment by tmakatos** [comment]
FYI Changpeng was using this branch: https://github.com/nutanix/libvfio-user/pull/457

**Comment by swapnili** [comment]
Please have another look. Also added a UT. 
I am gonna fix vfu_dma_read / vfu_dma_write() in separate patch.

**Review by jlevon — APPROVED** [review]
OK, please remember to file a bug on vfu_dma_read/write() as it affects the spec.

**Comment by swapnili** [comment]
https://github.com/nutanix/libvfio-user/issues/495

**Pull request merged**